grin/src/Reducer/LLVM/PrimOps.hs

```diff
@@ -13,66 +14,88 @@ import qualified Grin.Grin as Grin
 import Grin.TypeEnv hiding (function)
 import Reducer.LLVM.Base
 import Reducer.LLVM.TypeGen
+import Grin.PrimOpsPrelude
+
 
 cgUnit    = toCGType $ T_SimpleType T_Unit    :: CGType
 cgInt64   = toCGType $ T_SimpleType T_Int64   :: CGType
 cgWord64  = toCGType $ T_SimpleType T_Word64  :: CGType
 cgFloat   = toCGType $ T_SimpleType T_Float   :: CGType
 cgBool    = toCGType $ T_SimpleType T_Bool    :: CGType
+cgString  = toCGType $ T_SimpleType T_String  :: CGType
+cgChar    = toCGType $ T_SimpleType T_Char    :: CGType
+
+codeExternal :: Grin.External -> [Operand] -> CG (Either String Result)
+codeExternal e ops = case Grin.eKind e of
+  Grin.PrimOp -> codeGenPrimOp (Grin.eName e) ops
+  Grin.FFI    -> codeGenFFI     e             ops
 
-codeGenPrimOp :: Grin.Name -> [Grin.Val] -> [Operand] -> CG Result
-codeGenPrimOp name _ [opA, opB] = case name of
+codeGenPrimOp :: Grin.Name -> [Operand] -> CG (Either String Result)
```

Comment:
Why to use `Either` instead of `error`?

Reply:
Ah yes, this has no use after the second iteration of string handling, changing...